src/main/scala/cognite/spark/v1/DefaultSource.scala

```diff
@@ -188,104 +188,108 @@ class DefaultSource
-      val dataRepartitioned =
-        if (originalNumberOfPartitions > 50 && originalNumberOfPartitions > idealNumberOfPartitions) {
-          data.repartition(idealNumberOfPartitions)
-        } else {
-          data
-        }
-
-      dataRepartitioned.foreachPartition((rows: Iterator[Row]) => {
-        import CdpConnector.ioRuntime
 
-        val maxParallelism = config.parallelismPerPartition
-        val batches = Stream.fromIterator[IO](rows, chunkSize = batchSize).chunks
+    val relation: CdfRelation = resourceType match {
+      case AssetHierarchyBuilder.name =>
+        new AssetHierarchyBuilder(config)(sqlContext)
+      case NumericDataPointsRelation.name =>
+        new NumericDataPointsRelationV1(config)(sqlContext)
+      case StringDataPointsRelation.name =>
+        new StringDataPointsRelationV1(config)(sqlContext)
+      case EventsRelation.name =>
+        new EventsRelation(config)(sqlContext)
+      case TimeSeriesRelation.name =>
+        new TimeSeriesRelation(config)(sqlContext)
+      case AssetsRelation.name =>
+        new AssetsRelation(config)(sqlContext)
+      case FilesRelation.name =>
+        new FilesRelation(config)(sqlContext)
+      case SequenceRelation.name =>
+        new SequencesRelation(config)(sqlContext)
+      case LabelsRelation.name =>
+        new LabelsRelation(config)(sqlContext)
+      case SequenceRowsRelation.name =>
+        createSequenceRows(parameters, config, sqlContext)
+      case RelationshipsRelation.name =>
+        new RelationshipsRelation(config)(sqlContext)
+      case DataSetsRelation.name =>
+        new DataSetsRelation(config)(sqlContext)
+      case FlexibleDataModelBaseRelation.name =>
+        createFlexibleDataModelRelation(parameters, config, sqlContext)
+      case _ => sys.error(s"Resource type $resourceType does not support save()")
+    }
 
-        val operation: Seq[Row] => IO[Unit] = config.onConflict match {
-          case OnConflictOption.Abort =>
-            relation.insert
-          case OnConflictOption.Upsert =>
-            relation.upsert
-          case OnConflictOption.Update =>
-            relation.update
+    relation match {
+      case relation: DataPointsRelationV1[_] =>
+        if (config.onConflict == OnConflictOption.Delete) {
+          // Datapoints support 100_000 per request when inserting, but only 10_000 when deleting
+          val batchSize = config.batchSize.getOrElse(Constants.DefaultDataPointsLimit)
+          data.foreachPartition((rows: Iterator[Row]) => {
+            import CdpConnector.ioRuntime
+            val batches = rows.grouped(batchSize).toVector
+            batches.parTraverse_(relation.delete).unsafeRunSync()
+          })
+        } else {
+          // datapoints need special handling of dataframes and batches
+          relation.insert(data, overwrite = true)
+        }
+        relation
+      case relation: AssetHierarchyBuilder =>
```

Comment:
```suggestion
      case relation: AssetHierarchyBuilder with WritableRelation =>
```
looks like in spirit it should also be checked (is it valid Scala syntax?)

Reply:
It doesn't implement it, but maybe it should ^^

Reply:
Hmm, probably it should, I'll add it I guess

Reply:
oh, that's annoying, maybe we need a more fine-grained trait then to make it super clean

Reply:
ah it's even worse.

It implements insert/delete/update but the signature is different (doesn't take row, takes vector of AssetUpdateSchema/AssetInsertSchema etc...) room for improvement here. I was looking at fixing this but looks too annoying to be in the same PR at elast